File: leetcode/Partition_List.java
import java.util.*;

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode partition(ListNode head, int x) {
        ListNode head_a = null;
        ListNode head_b = null;
        ListNode tail_a = null;
        ListNode tail_b = null;
        ListNode cur = head;
        while (cur != null) {
            ListNode cur_next = cur.next;
            if (cur.val < x) {
                if (tail_a == null) {
                    head_a = tail_a = cur;
                } else {
                    tail_a.next = cur;
                    tail_a = cur;
                }
                tail_a.next = null;
            } else {
                if (tail_b == null) {
                    head_b = tail_b = cur;
                } else {
                    tail_b.next = cur;
                    tail_b = cur;
                }
                tail_b.next = null;
            }
            cur = cur_next;
        }
        if (head_a == null) {
            return head_b;
        }
        tail_a.next = head_b;
        return head_a;
    }
}

File: leetcode/Partition_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *partition(ListNode *head, int x) {
        ListNode *small_list_head = NULL;
        ListNode *small_list_tail = NULL;
        ListNode *large_list_head = NULL;
        ListNode *large_list_tail = NULL;
        ListNode *current = head;
        while (current != NULL) {
            if (current->val < x) {
                if (small_list_head == NULL) {
                    small_list_head = small_list_tail = current;
                } else {
                    small_list_tail->next = current;
                    small_list_tail = current;
                }
            } else {
                if (large_list_head == NULL) {
                    large_list_head = large_list_tail = current;
                } else {
                    large_list_tail->next = current;
                    large_list_tail = current;
                }
            }
            current = current->next;
        }
        if (small_list_head == NULL) {
            return large_list_head;
        } else {
            small_list_tail->next = large_list_head;
            if (large_list_tail != NULL)  // large_list might be empty
                large_list_tail->next = NULL;
            return small_list_head;
        }
    }
};
